File: clients/java/testcontainers/src/test/java/io/mockly/testcontainers/MocklyContainerIntegrationTest.java
package io.mockly.testcontainers;

import io.mockly.driver.model.CallEntry;
import io.mockly.driver.model.CallSummary;
import io.mockly.driver.model.Mock;
import io.mockly.driver.model.MockRequest;
import io.mockly.driver.model.MockResponse;
import io.mockly.driver.model.MockResponsePatch;
import io.mockly.driver.model.Scenario;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;
import org.testcontainers.junit.jupiter.Container;
import org.testcontainers.junit.jupiter.Testcontainers;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * Integration tests that spin up a real Mockly Docker container.
 *
 * <p>These tests require a running Docker daemon. Run with:
 * <pre>mvn test -Dgroups=integration</pre>
 *
 * <p>Excluded from the default build (no {@code integration} tag) via:
 * <pre>mvn test -Dgroups='!integration'</pre>
 */
@Tag("integration")
@Testcontainers
class MocklyContainerIntegrationTest {

    @Container
    private static final MocklyContainer container = new MocklyContainer();

    private final HttpClient http = HttpClient.newHttpClient();

    @BeforeEach
    void resetContainer() throws IOException, InterruptedException {
        container.reset();
    }

    @Test
    void containerStartsAndApiIsReachable() throws IOException, InterruptedException {
        String apiBase = container.getApiBase();
        assertTrue(apiBase.startsWith("http://"), "apiBase should start with http://");

        HttpResponse<String> resp = http.send(
                HttpRequest.newBuilder().uri(URI.create(apiBase + "/api/protocols")).GET().build(),
                HttpResponse.BodyHandlers.ofString());
        assertEquals(200, resp.statusCode(), "GET /api/protocols should return 200");
    }

    @Test
    void addMockAndHitEndpointThenReset() throws IOException, InterruptedException {
        Mock mock = Mock.builder(
                "hello-mock",
                MockRequest.builder("GET", "/hello").build(),
                MockResponse.builder(200).body("world").build()
        ).build();

        container.addMock(mock);

        HttpResponse<String> resp = http.send(
                HttpRequest.newBuilder().uri(URI.create(container.getHttpBase() + "/hello")).GET().build(),
                HttpResponse.BodyHandlers.ofString());
        assertEquals(200, resp.statusCode(), "mock should return 200");
        assertEquals("world", resp.body(), "mock should return 'world'");

        container.reset();

        HttpResponse<String> afterReset = http.send(
                HttpRequest.newBuilder().uri(URI.create(container.getHttpBase() + "/hello")).GET().build(),
                HttpResponse.BodyHandlers.ofString());
        assertNotEquals(200, afterReset.statusCode(), "mock should be cleared after reset");
    }

    @Test
    void getLogsReturnsEntriesAfterRequest() throws IOException, InterruptedException {
        http.send(
                HttpRequest.newBuilder().uri(URI.create(container.getHttpBase() + "/log-probe")).GET().build(),
                HttpResponse.BodyHandlers.ofString());

        List<CallEntry> logs = container.getApiLogs();
        assertFalse(logs.isEmpty(), "getApiLogs() should return parsed log entries");
        assertTrue(logs.stream().anyMatch(entry -> "/log-probe".equals(entry.getPath())),
                "expected /log-probe entry in logs");
    }

    @Test
    void withInlineConfigStartsContainer() {
        assertDoesNotThrow(() -> {
            String apiBase = container.getApiBase();
            HttpResponse<String> resp = http.send(
                    HttpRequest.newBuilder().uri(URI.create(apiBase + "/api/protocols")).GET().build(),
                    HttpResponse.BodyHandlers.ofString());
            assertEquals(200, resp.statusCode());
        });
    }

    @Test
    void listMocksReturnsAddedMocks() throws IOException, InterruptedException {
        Mock mock = Mock.builder(
                "list-test",
                MockRequest.builder("GET", "/list-test").build(),
                MockResponse.builder(200).body("listed").build()
        ).build();

        container.addMock(mock);

        List<Mock> mocks = container.listMocks();
        assertFalse(mocks.isEmpty(), "listMocks() should return added mocks");
        assertEquals("list-test", mocks.get(0).getId());
    }

    @Test
    void updateAndPatchMock() throws IOException, InterruptedException {
        container.addMock(Mock.builder(
                "upd-test",
                MockRequest.builder("GET", "/upd-test").build(),
                MockResponse.builder(200).body("orig").build()
        ).build());

        container.updateMock("upd-test", Mock.builder(
                "upd-test",
                MockRequest.builder("GET", "/upd-test").build(),
                MockResponse.builder(200).body("updated").build()
        ).build());

        HttpResponse<String> updated = http.send(
                HttpRequest.newBuilder().uri(URI.create(container.getHttpBase() + "/upd-test")).GET().build(),
                HttpResponse.BodyHandlers.ofString());
        assertEquals(200, updated.statusCode());
        assertEquals("updated", updated.body());

        container.patchMock("upd-test", MockResponsePatch.builder().status(418).build());

        HttpResponse<String> patched = http.send(
                HttpRequest.newBuilder().uri(URI.create(container.getHttpBase() + "/upd-test")).GET().build(),
                HttpResponse.BodyHandlers.ofString());
        assertEquals(418, patched.statusCode());
    }

    @Test
    void stateCrudWorks() throws IOException, InterruptedException {
        container.setState(Map.of("k", "v"));

        Map<String, String> state = container.getState();
        assertTrue(state.containsKey("k"));
        assertEquals("v", state.get("k"));

        container.deleteState("k");

        Map<String, String> stateAfterDelete = container.getState();
        assertFalse(stateAfterDelete.containsKey("k"));
    }

    @Test
    void getLogsCountPositiveAfterRequest() throws IOException, InterruptedException {
        http.send(
                HttpRequest.newBuilder().uri(URI.create(container.getHttpBase() + "/count-probe")).GET().build(),
                HttpResponse.BodyHandlers.ofString());

        assertTrue(container.getLogsCount() > 0, "getLogsCount() should be positive after a request");
        assertDoesNotThrow(() -> container.getLogsCount("nonexistent"));
    }

    @Test
    void scenarioCrudWorks() throws IOException, InterruptedException {
        assertDoesNotThrow(() -> container.listScenarios());

        container.createScenario(Scenario.builder("tc-java-s", "TC Java").build());

        Scenario scenario = container.getScenario("tc-java-s");
        assertEquals("tc-java-s", scenario.getId());
        assertFalse(container.listScenarios().isEmpty(), "listScenarios() should contain created scenario");

        container.deleteScenario("tc-java-s");
    }

    @Test
    void getCallsAndClearWorks() throws IOException, InterruptedException {
        container.addMock(Mock.builder(
                "java-calls",
                MockRequest.builder("GET", "/java-calls").build(),
                MockResponse.builder(200).body("ok").build()
        ).build());

        http.send(
                HttpRequest.newBuilder().uri(URI.create(container.getHttpBase() + "/java-calls")).GET().build(),
                HttpResponse.BodyHandlers.ofString());

        CallSummary calls = container.getCalls("java-calls");
        assertTrue(calls.getCount() > 0, "getCalls() should report recorded calls");

        container.clearCalls("java-calls");

        CallSummary clearedCalls = container.getCalls("java-calls");
        assertEquals(0L, clearedCalls.getCount());

        assertDoesNotThrow(() -> container.clearAllCalls());
    }

    @Test
    void waitForCallsResolvesWhenHit() throws IOException, InterruptedException {
        container.addMock(Mock.builder(
                "java-wait",
                MockRequest.builder("GET", "/java-wait").build(),
                MockResponse.builder(200).body("ok").build()
        ).build());

        http.send(
                HttpRequest.newBuilder().uri(URI.create(container.getHttpBase() + "/java-wait")).GET().build(),
                HttpResponse.BodyHandlers.ofString());

        CallSummary summary = container.waitForCalls("java-wait", 1, Duration.ofSeconds(10));
        assertTrue(summary.getCount() >= 1, "waitForCalls() should observe at least one call");
    }
}

File: clients/node-testcontainers/src/MocklyContainer.integration.test.ts
import { afterAll, beforeAll, beforeEach, describe, expect, test } from 'vitest'

import { MocklyContainerBuilder, type StartedMocklyContainer } from './index.js'

const describeIntegration = process.env.INTEGRATION ? describe : describe.skip

const DEFAULT_INLINE_CONFIG = `mockly:
  api:
    port: 9091
protocols:
  http:
    enabled: true
    port: 8090
`

describeIntegration('MocklyContainer integration', () => {
  let container: StartedMocklyContainer

  beforeAll(async () => {
    container = await new MocklyContainerBuilder().start()
  })

  beforeEach(async () => {
    await container.reset()
    await container.clearLogs()
    await container.clearAllCalls()
  })

  afterAll(async () => {
    await container.stop()
  })

  test('container lifecycle and mock API', async () => {
    expect(container.getHttpBase()).toMatch(/^http:\/\//)
    expect(container.getApiBase()).toMatch(/^http:\/\//)

    const protocolsResponse = await fetch(`${container.getApiBase()}/api/protocols`)
    expect(protocolsResponse.status).toBe(200)

    await container.addMock({
      id: 'hello-mock',
      request: { method: 'GET', path: '/hello' },
      response: { status: 200, body: 'world' },
    })

    const mockResponse = await fetch(`${container.getHttpBase()}/hello`)
    expect(mockResponse.status).toBe(200)
    expect(await mockResponse.text()).toBe('world')

    await container.reset()

    const resetResponse = await fetch(`${container.getHttpBase()}/hello`)
    expect(resetResponse.status).not.toBe(200)
  })

  test('getLogs returns parsed call entries after request', async () => {
    await fetch(`${container.getHttpBase()}/any-path`)

    const logs = await container.getLogs()

    expect(logs.length).toBeGreaterThan(0)
    expect(logs[0]).toHaveProperty('path')
  })

  test('listMocks returns added mocks', async () => {
    await container.addMock({
      id: 'list-test',
      request: { method: 'GET', path: '/list-me' },
      response: { status: 200, body: 'listed' },
    })

    const mocks = await container.listMocks()

    expect(mocks.length).toBeGreaterThan(0)
    expect(mocks.some((mock) => mock.id === 'list-test')).toBe(true)
  })

  test('updateMock changes response', async () => {
    await container.addMock({
      id: 'update-test',
      request: { method: 'GET', path: '/update-me' },
      response: { status: 200, body: 'original' },
    })

    await container.updateMock('update-test', {
      id: 'update-test',
      request: { method: 'GET', path: '/update-me' },
      response: { status: 200, body: 'updated' },
    })

    const response = await fetch(`${container.getHttpBase()}/update-me`)

    expect(await response.text()).toBe('updated')
  })

  test('patchMock changes status', async () => {
    await container.addMock({
      id: 'patch-test',
      request: { method: 'GET', path: '/patch-me' },
      response: { status: 200, body: 'ok' },
    })

    await container.patchMock('patch-test', { status: 418 })

    const response = await fetch(`${container.getHttpBase()}/patch-me`)

    expect(response.status).toBe(418)
  })

  test('state CRUD works', async () => {
    await container.setState({ foo: 'bar' })

    const stateAfterSet = await container.getState()
    expect(stateAfterSet.foo).toBe('bar')

    await container.deleteState('foo')

    const stateAfterDelete = await container.getState()
    expect(stateAfterDelete.foo).toBeUndefined()
  })

  test('getLogsCount returns positive count after request', async () => {
    await fetch(container.getHttpBase())

    const count = await container.getLogsCount()

    expect(count).toBeGreaterThan(0)
  })

  test('scenario CRUD works', async () => {
    await expect(container.listScenarios()).resolves.toBeDefined()

    await container.createScenario({
      id: 'tc-s1',
      name: 'TC S1',
      patches: [],
    })

    const scenario = await container.getScenario('tc-s1')
    expect(scenario.id).toBe('tc-s1')

    const scenarios = await container.listScenarios()
    expect(scenarios.length).toBeGreaterThan(0)

    await container.deleteScenario('tc-s1')
  })

  test('getCalls and clearCalls work', async () => {
    await container.addMock({
      id: 'calls-test',
      request: { method: 'GET', path: '/calls-test' },
      response: { status: 200, body: 'ok' },
    })

    await fetch(`${container.getHttpBase()}/calls-test`)

    const callsAfterHit = await container.getCalls('calls-test')
    expect(callsAfterHit.count).toBeGreaterThan(0)

    await container.clearCalls('calls-test')

    const callsAfterClear = await container.getCalls('calls-test')
    expect(callsAfterClear.count).toBe(0)

    await expect(container.clearAllCalls()).resolves.toBeUndefined()
  })

  test('waitForCalls resolves when mock is hit', async () => {
    await container.addMock({
      id: 'wait-test',
      request: { method: 'GET', path: '/wait-test' },
      response: { status: 200, body: 'ok' },
    })

    const requestPromise = fetch(`${container.getHttpBase()}/wait-test`)
    const calls = await container.waitForCalls('wait-test', 1, '10s')

    await requestPromise

    expect(calls.count).toBeGreaterThanOrEqual(1)
  })
})

describeIntegration('MocklyContainer inline config integration', () => {
  let container: StartedMocklyContainer

  beforeAll(async () => {
    container = await new MocklyContainerBuilder()
      .withInlineConfig(DEFAULT_INLINE_CONFIG)
      .start()
  })

  afterAll(async () => {
    await container.stop()
  })

  test('withInlineConfig works', async () => {
    const response = await fetch(`${container.getApiBase()}/api/protocols`)
    expect(response.status).toBe(200)
  })
})

describeIntegration('MocklyContainer options integration', () => {
  let container: StartedMocklyContainer

  beforeAll(async () => {
    container = await new MocklyContainerBuilder()
      .withOptions({
        scenarios: [
          {
            id: 'payments-down',
            name: 'Payments down',
            patches: [{ mock_id: 'charge', status: 503 }],
          },
        ],
      })
      .start()
  })

  afterAll(async () => {
    await container.stop()
  })

  test('withOptions preloads scenarios', async () => {
    const scenarios = await container.listScenarios()
    expect(scenarios.map((scenario) => scenario.id)).toContain('payments-down')
  })
})
